spikeinterface/core/waveform_tools.py

```diff
@@ -0,0 +1,141 @@
+
+import numpy as np
+
+from .job_tools import ChunkRecordingExecutor, ensure_n_jobs, _shared_job_kwargs_doc
+
+def allocate_waveforms(recording, spikes, unit_ids, nbefore, nafter, mode='memmap', folder=None, dtype=None, sparsity_mask=None):
+    """
+    
+    
+    sparsity_mask:
+        If not None must
+        shape=(len(unit_ids), len(channel_ids)
+        dtype='bool'
+        
+    
+    """
+    
+    nsamples = nbefore + nafter
+    
+    if mode =='memmap':
+        # prepare memmap
+        wfs_arrays = {}
+        for unit_ind, unit_id in enumerate(unit_ids):
+            n_spikes = np.sum(spikes['unit_ind'] == unit_ind)
+            if sparsity_mask is None:
+                num_chans = recording.get_num_channels()
+            else:
+                num_chans = np.sum(sparsity_mask[unit_ind, :])
+            shape = (n_spikes, nsamples, num_chans)
+            filename = folder / f'waveforms_{unit_id}.npy'
+            arr = np.lib.format.open_memmap(filename, mode='w+', dtype=dtype, shape=shape)
+            wfs_arrays[unit_id] = arr
+        return wfs_arrays
+    elif mode =='memory':
+        raise NotImplementedError()
+    else:
+        raise ValueError('allocate_waveforms bad mode')
+
+
+def distribute_waveform_to_buffers(recording, spikes, unit_ids, wfs_arrays, nbefore, nafter, return_scaled, sparsity_mask=None, **job_kwargs):
+    n_jobs = ensure_n_jobs(recording, job_kwargs.get('n_jobs', None))
+
+    inds_by_unit = {}
+    for unit_ind, unit_id in enumerate(unit_ids):
+        inds,  = np.nonzero(spikes['unit_ind'] == unit_ind)
+        inds_by_unit[unit_id] = inds
+
+    # and run
+    func = _waveform_extractor_chunk
+    init_func = _init_worker_waveform_extractor
+    if n_jobs == 1:
+        init_args = (recording, )
+    else:
+        init_args = (recording.to_dict(), )
+    init_args = init_args + (unit_ids, spikes, wfs_arrays, nbefore, nafter, return_scaled, inds_by_unit, sparsity_mask)
+    processor = ChunkRecordingExecutor(recording, func, init_func, init_args, job_name='extract waveforms',
+                                       **job_kwargs)
+    processor.run()
+
+
+# used by WaveformExtractor + ChunkRecordingExecutor
+def _init_worker_waveform_extractor(recording, unit_ids, spikes, wfs_arrays, nbefore, nafter, return_scaled, inds_by_unit, sparsity_mask):
+    # create a local dict per worker
+    worker_ctx = {}
+    if isinstance(recording, dict):
+        from spikeinterface.core import load_extractor
+        recording = load_extractor(recording)
+    worker_ctx['recording'] = recording
+    
```

Comment:
@samuelgarcia I think you might need to reopen the memmap in each worker!

Reply:
I known. But I thhink I will do this only for windows and osx. So on linux it will avoid to much open files.